kong/router/atc_compat.lua

```diff
@@ -0,0 +1,687 @@
+  if req_method and type(req_method) ~= "string" then
+    error("method must be a string", 2)
+  end
+  if req_uri and type(req_uri) ~= "string" then
+    error("uri must be a string", 2)
+  end
+  if req_host and type(req_host) ~= "string" then
+    error("host must be a string", 2)
+  end
+  if req_scheme and type(req_scheme) ~= "string" then
+    error("scheme must be a string", 2)
+  end
+  if _src_ip and type(_src_ip) ~= "string" then
+    error("src_ip must be a string", 2)
+  end
+  if _src_port and type(_src_port) ~= "number" then
+    error("src_port must be a number", 2)
+  end
+  if _dst_ip and type(_dst_ip) ~= "string" then
+    error("dst_ip must be a string", 2)
+  end
+  if _dst_port and type(_dst_port) ~= "number" then
+    error("dst_port must be a number", 2)
+  end
+  if sni and type(sni) ~= "string" then
+    error("sni must be a string", 2)
+  end
+  if req_headers and type(req_headers) ~= "table" then
+    error("headers must be a table", 2)
+  end
+
+  local c = context.new(self.schema)
+
+  for _, field in ipairs(self.fields) do
+    if field == "http.method" then
+      assert(c:add_value("http.method", req_method))
+
+    elseif field == "http.path" then
+      assert(c:add_value("http.path", req_uri))
+
+    elseif field == "http.host" then
+      assert(c:add_value("http.host", req_host))
+
+    elseif field == "net.protocol" then
+      assert(c:add_value("net.protocol", req_scheme))
+
+    elseif field == "tls.sni" then
+      assert(c:add_value("tls.sni", sni))
+
+    elseif req_headers and field:sub(1, 13) == "http.headers." then
+      local h = field:sub(14)
+      local v = req_headers[h]
+
+      if v then
+        if type(v) == "string" then
+          assert(c:add_value(field, v:lower()))
+
+        else
+          for _, v in ipairs(v) do
+            assert(c:add_value(field, v:lower()))
+          end
+        end
+      end
+    end
+  end
+
+  local matched = self.router:execute(c)
+  if not matched then
+    return nil
+  end
+
+  local uuid, matched_path, captures = c:get_result("http.path")
+
+  local service_protocol
+  local service_type
+  local service_host
+  local service_port
+  local service = self.services[uuid]
+  local matched_route = self.routes[uuid]
+
+  if service then
+    service_protocol = service.protocol
+    service_host = service.host
+    service_port = service.port
+  end
+
+  if service_protocol then
+    service_type = protocol_subsystem[service_protocol]
+  end
+
+  local service_hostname_type
+  if service_host then
+    service_hostname_type = hostname_type(service_host)
+  end
+
+  if not service_port then
+    if service_protocol == "https" then
+      service_port = 443
+    elseif service_protocol == "http" then
+      service_port = 80
+    end
+  end
+
+  local service_path
+  if service_type == "http" then
+    service_path = service and service.path or "/"
+  end
+
+  local request_prefix = matched_route.strip_path and matched_path or nil
```

Comment:
556-608 is a huge code block, if possible we should move them into a function.

Reply:
This was copied directly from the old router. We can refactor it to a helper function. I added a TODO in the code for now.